File: queen_proxy/scripts/sdk-smoke/js/smoke.mjs
#!/usr/bin/env node
/**
 * sdk-smoke (JavaScript) -- the REAL published client (clients/client-js,
 * package `queen-mq`, entry point client-v2/index.js) driven against a live
 * queen_proxy + broker. No canned HTTP server anywhere in this file.
 *
 * One phase per invocation (argv[1]); the driver (../run.sh) owns the cell and
 * the control plane, this program owns the SDK. Output mirrors
 * scripts/isolation-smoke.sh: "  ok  - desc" / "  FAIL- desc", exit 1 on any
 * failure so the driver can score a language per phase.
 */
import { Queen } from '../../../../clients/client-js/client-v2/index.js'

const PHASE = process.argv[2] || ''

// A queen_proxy deployment selects the cluster from the Host header, and there
// are two ways a client can produce one:
//   (a) point the base URL at the cluster's own hostname (https://two.<zone>) --
//       what production DNS gives every customer;
//   (b) keep one base URL and set `hostHeader: '<slug>'` -- the cell address is
//       dialled while the request advertises the cluster, which is what makes
//       this runnable against 127.0.0.1 with no wildcard DNS.
// This harness uses (b). `headers: { Host }` also works (all three clients map
// it onto the supported mechanism); the isolation phase asserts that parity,
// since Node's fetch would otherwise drop the header silently.

function env(name, dflt) {
  const v = process.env[name] ?? dflt
  if (v === undefined) {
    console.error(`sdk-smoke(js): missing env ${name}`)
    process.exit(2)
  }
  return v
}

const URL_ = env('SDK_SMOKE_URL')
const HOST_A = env('SDK_SMOKE_HOST_A')
const KEY_A = env('SDK_SMOKE_KEY_A')
const HOST_B = env('SDK_SMOKE_HOST_B')
const KEY_B = env('SDK_SMOKE_KEY_B')
const QUEUE = env('SDK_SMOKE_QUEUE')
const ISO_QUEUE = env('SDK_SMOKE_ISO_QUEUE')
const RUN_ID = env('SDK_SMOKE_RUN_ID')
const DEADLINE_MS = Number(env('SDK_SMOKE_DEADLINE_MS', '150000'))
const BURN_MAX = Number(env('SDK_SMOKE_BURN_MAX', '400'))
const RECOVER_N = Number(env('SDK_SMOKE_RECOVER_N', '20'))

let PASS = 0
let FAIL = 0
const ok = (d) => { PASS++; console.log(`  ok  - ${d}`) }
const bad = (d) => { FAIL++; console.log(`  FAIL- ${d}`) }
const check = (desc, cond, detail) => cond ? ok(desc) : bad(detail ? `${desc} (${detail})` : desc)
const sleep = (ms) => new Promise(r => setTimeout(r, ms))

/** A client wired the way a customer would wire one at a proxy: the cell's
 *  address plus the cluster's advertised hostname, which is what selects the
 *  tenant. `hostHeader` keeps the socket on the cell while the request
 *  advertises the cluster — the same shape as curl --resolve, and the reason
 *  this needs no wildcard DNS locally. */
function client(slug, key, extra = {}) {
  return new Queen({
    url: URL_,
    hostHeader: slug,
    bearerToken: key,
    handleSignals: false,
    ...extra,
  })
}

// --------------------------------------------------------------------------
// 1. push -> pop -> ack round trip through the proxy, api-key auth
// --------------------------------------------------------------------------
async function roundtrip() {
  const q = client(HOST_A, KEY_A)
  try {
    const res = await q.queue(QUEUE).push([
      { n: 1, run: RUN_ID }, { n: 2, run: RUN_ID }, { n: 3, run: RUN_ID },
    ])
    check('push 3 items accepted through the proxy',
      Array.isArray(res) && res.length === 3 && res.every(r => r.status === 'queued'),
      JSON.stringify(res))

    const msgs = await q.queue(QUEUE).batch(10).wait(false).pop()
    const ns = msgs.map(m => m.data && m.data.n).sort()
    check('pop returns the 3 pushed messages',
      msgs.length === 3 && JSON.stringify(ns) === '[1,2,3]',
      `got ${msgs.length}: ${JSON.stringify(ns)}`)
    check('messages carry partitionId + transactionId',
      msgs.length > 0 && msgs.every(m => m.partitionId && m.transactionId))

    await q.ack(msgs, true)
    const after = await q.queue(QUEUE).batch(10).wait(false).pop()
    check('queue drained after ack', after.length === 0, `got ${after.length}`)
  } finally {
    await q.close()
  }
}

// --------------------------------------------------------------------------
// 2. tenant isolation with the real client: same queue name, two clusters
// --------------------------------------------------------------------------
async function isolation() {
  const a = client(HOST_A, KEY_A)
  const b = client(HOST_B, KEY_B)
  const foreign = client(HOST_B, KEY_A) // A's key presented on B's Host
  try {
    const ra = await a.queue(ISO_QUEUE).push([{ who: 'A', run: RUN_ID }])
    const rb = await b.queue(ISO_QUEUE).push([{ who: 'B', run: RUN_ID }])
    check('both clusters accept a push to the same queue name',
      ra[0].status === 'queued' && rb[0].status === 'queued',
      `${ra[0].status}/${rb[0].status}`)

    const ma = await a.queue(ISO_QUEUE).batch(10).wait(false).pop()
    const mb = await b.queue(ISO_QUEUE).batch(10).wait(false).pop()
    check('cluster A sees only its own message',
      ma.length === 1 && ma[0].data.who === 'A',
      `${ma.length} msg(s): ${JSON.stringify(ma.map(m => m.data))}`)
    check('cluster B sees only its own message',
      mb.length === 1 && mb[0].data.who === 'B',
      `${mb.length} msg(s): ${JSON.stringify(mb.map(m => m.data))}`)

    let crossErr = null
    try {
      await foreign.queue(ISO_QUEUE).push([{ who: 'X', run: RUN_ID }])
    } catch (e) {
      crossErr = e
    }
    check('cluster A key on cluster B host -> 403 forbidden',
      crossErr !== null && crossErr.status === 403 && crossErr.code === 'forbidden',
      crossErr ? `${crossErr.status}/${crossErr.code}` : 'push SUCCEEDED')

    // Cross-language parity on the other documented way to select a cluster:
    // one base URL + a Host header. Go maps it onto req.Host, httpx sends it
    // verbatim, and the JS client maps it onto hostHeader (Node's fetch would
    // otherwise drop it silently and land the request on the wrong cluster).
    // Kept as an assertion because a silent drop here is a wrong-tenant bug.
    const byHeader = new Queen({
      url: URL_, bearerToken: KEY_B, headers: { Host: HOST_B }, handleSignals: false,
    })
    let hdrRes = null
    let hdrErr = null
    try {
      hdrRes = await byHeader.queue(ISO_QUEUE).push([{ who: 'B-hdr', run: RUN_ID }])
    } catch (e) {
      hdrErr = e
    } finally {
      await byHeader.close()
    }
    check("headers:{ Host } selects the cluster (parity with Go/Python)",
      hdrRes !== null && hdrRes[0] && hdrRes[0].status === 'queued',
      hdrErr ? `${hdrErr.status}/${hdrErr.code}: ${hdrErr.message}` : 'no response')

    await a.ack(ma, true)
    await b.ack(mb, true)
    // The parity push above landed a real message on B (it is only skippable
    // when the header mechanism is broken), so B has one more to drain than A.
    const extra = await b.queue(ISO_QUEUE).batch(10).wait(false).pop()
    check('the Host-header push landed on B, not A',
      extra.length === 1 && extra[0].data.who === 'B-hdr',
      `${extra.length} msg(s): ${JSON.stringify(extra.map(m => m.data))}`)
    await b.ack(extra, true)

    const ea = await a.queue(ISO_QUEUE).batch(10).wait(false).pop()
    const eb = await b.queue(ISO_QUEUE).batch(10).wait(false).pop()
    check('both clusters drained independently after ack',
      ea.length === 0 && eb.length === 0, `${ea.length}/${eb.length}`)
  } finally {
    await Promise.all([a.close(), b.close(), foreign.close()])
  }
}

// --------------------------------------------------------------------------
// 3. live 429 from the real limiter + transparent recovery
// --------------------------------------------------------------------------
async function ratelimit() {
  // (a) a client with 429 backoff DISABLED, so the live 429 surfaces as an
  //     error we can inspect (code + Retry-After) instead of being absorbed.
  const strict = client(HOST_A, KEY_A, { retry429: { maxAttempts: 1 } })
  let hit = null
  let sent = 0
  try {
    for (let i = 0; i < BURN_MAX && !hit; i++) {
      try {
        await strict.queue(QUEUE).push([{ burn: i, run: RUN_ID }])
        sent++
      } catch (e) {
        if (e.status === 429) { hit = e; break }
        throw e
      }
    }
  } finally {
    await strict.close()
  }
  check('the real limiter returned a live 429 to the SDK', hit !== null,
    `no 429 after ${sent} pushes`)
  check('429 body carries code=rate_limited', hit !== null && hit.code === 'rate_limited',
    hit ? String(hit.code) : 'n/a')
  check('429 carries a Retry-After the client parsed',
    hit !== null && typeof hit.retryAfterSeconds === 'number' && hit.retryAfterSeconds >= 1,
    hit ? String(hit.retryAfterSeconds) : 'n/a')

  // (b) the same traffic through a stock client (default 429 policy): the
  //     bucket is empty, so every one of these has to be paced by the
  //     client's own Retry-After backoff -- and all of them must still land.
  const normal = client(HOST_A, KEY_A)
  const t0 = Date.now()
  let done = 0
  try {
    for (let i = 0; i < RECOVER_N; i++) {
      const r = await normal.queue(QUEUE).push([{ recover: i, run: RUN_ID }])
      if (r[0] && r[0].status === 'queued') done++
    }
  } finally {
    await normal.close()
  }
  const elapsed = Date.now() - t0
  check(`stock client completed all ${RECOVER_N} pushes against an empty bucket`,
    done === RECOVER_N, `${done}/${RECOVER_N}`)
  check('the run was paced by backoff, not served instantly', elapsed >= 1500, `${elapsed}ms`)
}

// --------------------------------------------------------------------------
// 4a. terminal 403: storage quota tripped by the driver's limit override
// --------------------------------------------------------------------------
async function blocked() {
  const q = client(HOST_A, KEY_A)
  const deadline = Date.now() + DEADLINE_MS
  let terminal = null
  let callMs = 0
  try {
    while (Date.now() < deadline && terminal === null) {
      const t = Date.now()
      try {
        await q.queue(QUEUE).push([{ probe: Date.now(), run: RUN_ID }])
      } catch (e) {
        callMs = Date.now() - t
        if (e.status === 403) { terminal = e; break }
        throw e
      }
      await sleep(2000)
    }
    check('push eventually rejected with a terminal 403', terminal !== null,
      `still accepted after ${Math.round(DEADLINE_MS / 1000)}s`)
    check('terminal code is storage_quota_exceeded',
      terminal !== null && terminal.code === 'storage_quota_exceeded',
      terminal ? String(terminal.code) : 'n/a')
    check('terminal 403 surfaced immediately (not retried with backoff)',
      terminal !== null && callMs < 2000, `${callMs}ms`)

    // consume must stay open while pushes are blocked: the rate-limit phase
    // left plenty of un-popped messages on this queue, so an empty result
    // here would mean the read path was blocked too.
    const msgs = await q.queue(QUEUE).batch(1).wait(false).pop()
    check('consume still allowed while push-blocked', msgs.length >= 1, `${msgs.length} msg(s)`)
    if (msgs.length) await q.ack(msgs, true)
  } finally {
    await q.close()
  }
}

// --------------------------------------------------------------------------
// 4b. recovery once the driver clears the override
// --------------------------------------------------------------------------
async function unblocked() {
  const q = client(HOST_A, KEY_A)
  const deadline = Date.now() + DEADLINE_MS
  let released = false
  let lastCode = null
  try {
    while (Date.now() < deadline && !released) {
      try {
        const r = await q.queue(QUEUE).push([{ release: Date.now(), run: RUN_ID }])
        released = Array.isArray(r) && r[0] && r[0].status === 'queued'
      } catch (e) {
        if (e.status !== 403) throw e
        lastCode = e.code
      }
      if (!released) await sleep(2000)
    }
    check('push accepted again after the override is cleared', released,
      `still ${lastCode} after ${Math.round(DEADLINE_MS / 1000)}s`)
  } finally {
    await q.close()
  }
}

const PHASES = { roundtrip, isolation, ratelimit, blocked, unblocked }

const fn = PHASES[PHASE]
if (!fn) {
  console.error(`sdk-smoke(js): unknown phase '${PHASE}' (want: ${Object.keys(PHASES).join('|')})`)
  process.exit(2)
}

try {
  await fn()
} catch (e) {
  bad(`phase threw: ${e && e.message} (status=${e && e.status} code=${e && e.code})`)
}
console.log(`  -- js/${PHASE}: ${PASS} ok, ${FAIL} fail`)
process.exit(FAIL === 0 ? 0 : 1)

File: proxy/scripts/sdk-smoke/py/smoke.py
#!/usr/bin/env python3
"""sdk-smoke (Python) -- the REAL published client (clients/client-py, package
`queen-mq`) driven against a live queen_proxy + broker. No MockTransport, no
canned server: every request in this file leaves the process.

One phase per invocation (argv[1]); the driver (../run.sh) owns the cell and
the control plane, this program owns the SDK. Output mirrors
scripts/isolation-smoke.sh: "  ok  - desc" / "  FAIL- desc", exit 1 on any
failure so the driver can score a language per phase.
"""

import asyncio
import json
import os
import sys
import time

HERE = os.path.dirname(os.path.abspath(__file__))
# ../../../../clients/client-py -- import the checked-out client by path, the
# same source a `pip install queen-mq` would ship. No publishing step.
CLIENT_PY = os.path.abspath(os.path.join(HERE, "..", "..", "..", "..", "clients", "client-py"))
sys.path.insert(0, CLIENT_PY)

import httpx  # noqa: E402  (the client's own HTTP dependency)
from queen import Queen  # noqa: E402


def env(name, default=None):
    v = os.environ.get(name, default)
    if v is None:
        print(f"sdk-smoke(py): missing env {name}", file=sys.stderr)
        sys.exit(2)
    return v


PHASE = sys.argv[1] if len(sys.argv) > 1 else ""
URL = env("SDK_SMOKE_URL")
HOST_A = env("SDK_SMOKE_HOST_A")
KEY_A = env("SDK_SMOKE_KEY_A")
HOST_B = env("SDK_SMOKE_HOST_B")
KEY_B = env("SDK_SMOKE_KEY_B")
QUEUE = env("SDK_SMOKE_QUEUE")
ISO_QUEUE = env("SDK_SMOKE_ISO_QUEUE")
RUN_ID = env("SDK_SMOKE_RUN_ID")
DEADLINE_S = float(env("SDK_SMOKE_DEADLINE_MS", "150000")) / 1000.0
BURN_MAX = int(env("SDK_SMOKE_BURN_MAX", "400"))
RECOVER_N = int(env("SDK_SMOKE_RECOVER_N", "20"))

STATE = {"pass": 0, "fail": 0}


def ok(desc):
    STATE["pass"] += 1
    print(f"  ok  - {desc}")


def bad(desc):
    STATE["fail"] += 1
    print(f"  FAIL- {desc}")


def check(desc, cond, detail=""):
    if cond:
        ok(desc)
    else:
        bad(f"{desc} ({detail})" if detail else desc)


def client(host, key, **extra):
    """A client wired exactly like a customer would wire one at a proxy: base
    URL, api key as bearer token, Host header selecting the cluster."""
    return Queen(url=URL, bearer_token=key, headers={"Host": host}, **extra)


def status_of(err):
    resp = getattr(err, "response", None)
    return getattr(resp, "status_code", None)


# ---------------------------------------------------------------------------
# 1. push -> pop -> ack round trip through the proxy, api-key auth
# ---------------------------------------------------------------------------
async def roundtrip():
    q = client(HOST_A, KEY_A)
    try:
        res = await q.queue(QUEUE).push(
            [{"n": 1, "run": RUN_ID}, {"n": 2, "run": RUN_ID}, {"n": 3, "run": RUN_ID}]
        )
        check(
            "push 3 items accepted through the proxy",
            isinstance(res, list) and len(res) == 3 and all(r.get("status") == "queued" for r in res),
            json.dumps(res)[:200],
        )

        msgs = await q.queue(QUEUE).batch(10).wait(False).pop()
        ns = sorted(m.get("data", {}).get("n") for m in msgs)
        check(
            "pop returns the 3 pushed messages",
            len(msgs) == 3 and ns == [1, 2, 3],
            f"got {len(msgs)}: {ns}",
        )
        check(
            "messages carry partitionId + transactionId",
            len(msgs) > 0 and all(m.get("partitionId") and m.get("transactionId") for m in msgs),
        )

        await q.ack(msgs, True)
        after = await q.queue(QUEUE).batch(10).wait(False).pop()
        check("queue drained after ack", len(after) == 0, f"got {len(after)}")
    finally:
        await q.close()


# ---------------------------------------------------------------------------
# 2. tenant isolation with the real client: same queue name, two clusters
# ---------------------------------------------------------------------------
async def isolation():
    a = client(HOST_A, KEY_A)
    b = client(HOST_B, KEY_B)
    foreign = client(HOST_B, KEY_A)  # A's key presented on B's Host
    try:
        ra = await a.queue(ISO_QUEUE).push([{"who": "A", "run": RUN_ID}])
        rb = await b.queue(ISO_QUEUE).push([{"who": "B", "run": RUN_ID}])
        check(
            "both clusters accept a push to the same queue name",
            ra[0].get("status") == "queued" and rb[0].get("status") == "queued",
            f"{ra[0].get('status')}/{rb[0].get('status')}",
        )

        ma = await a.queue(ISO_QUEUE).batch(10).wait(False).pop()
        mb = await b.queue(ISO_QUEUE).batch(10).wait(False).pop()
        check(
            "cluster A sees only its own message",
            len(ma) == 1 and ma[0]["data"]["who"] == "A",
            f"{len(ma)} msg(s): {[m.get('data') for m in ma]}",
        )
        check(
            "cluster B sees only its own message",
            len(mb) == 1 and mb[0]["data"]["who"] == "B",
            f"{len(mb)} msg(s): {[m.get('data') for m in mb]}",
        )

        cross = None
        try:
            await foreign.queue(ISO_QUEUE).push([{"who": "X", "run": RUN_ID}])
        except httpx.HTTPStatusError as e:
            cross = e
        check(
            "cluster A key on cluster B host -> 403 forbidden",
            cross is not None and status_of(cross) == 403 and getattr(cross, "code", None) == "forbidden",
            f"{status_of(cross)}/{getattr(cross, 'code', None)}" if cross else "push SUCCEEDED",
        )

        await a.ack(ma, True)
        await b.ack(mb, True)
        ea = await a.queue(ISO_QUEUE).batch(10).wait(False).pop()
        eb = await b.queue(ISO_QUEUE).batch(10).wait(False).pop()
        check(
            "both clusters drained independently after ack",
            len(ea) == 0 and len(eb) == 0,
            f"{len(ea)}/{len(eb)}",
        )
    finally:
        await a.close()
        await b.close()
        await foreign.close()


# ---------------------------------------------------------------------------
# 3. live 429 from the real limiter + transparent recovery
# ---------------------------------------------------------------------------
async def ratelimit():
    # (a) 429 backoff DISABLED, so the live 429 surfaces as an error we can
    #     inspect (code + Retry-After) instead of being absorbed.
    strict = client(HOST_A, KEY_A, retry_429={"max_attempts": 1})
    hit = None
    sent = 0
    try:
        for i in range(BURN_MAX):
            try:
                await strict.queue(QUEUE).push([{"burn": i, "run": RUN_ID}])
                sent += 1
            except httpx.HTTPStatusError as e:
                if status_of(e) == 429:
                    hit = e
                    break
                raise
    finally:
        await strict.close()

    check("the real limiter returned a live 429 to the SDK", hit is not None, f"no 429 after {sent} pushes")
    check(
        "429 body carries code=rate_limited",
        hit is not None and getattr(hit, "code", None) == "rate_limited",
        str(getattr(hit, "code", None)) if hit else "n/a",
    )
    ra = getattr(hit, "retry_after_seconds", None) if hit else None
    check(
        "429 carries a Retry-After the client parsed",
        isinstance(ra, float) and ra >= 1,
        str(ra),
    )

    # (b) the same traffic through a stock client (default 429 policy): the
    #     bucket is empty, so every one of these has to be paced by the
    #     client's own Retry-After backoff -- and all of them must still land.
    normal = client(HOST_A, KEY_A)
    t0 = time.monotonic()
    done = 0
    try:
        for i in range(RECOVER_N):
            r = await normal.queue(QUEUE).push([{"recover": i, "run": RUN_ID}])
            if r and r[0].get("status") == "queued":
                done += 1
    finally:
        await normal.close()
    elapsed_ms = int((time.monotonic() - t0) * 1000)
    check(
        f"stock client completed all {RECOVER_N} pushes against an empty bucket",
        done == RECOVER_N,
        f"{done}/{RECOVER_N}",
    )
    check("the run was paced by backoff, not served instantly", elapsed_ms >= 1500, f"{elapsed_ms}ms")


# ---------------------------------------------------------------------------
# 4a. terminal 403: storage quota tripped by the driver's limit override
# ---------------------------------------------------------------------------
async def blocked():
    q = client(HOST_A, KEY_A)
    deadline = time.monotonic() + DEADLINE_S
    terminal = None
    call_ms = 0
    try:
        while time.monotonic() < deadline and terminal is None:
            t = time.monotonic()
            try:
                await q.queue(QUEUE).push([{"probe": time.time(), "run": RUN_ID}])
            except httpx.HTTPStatusError as e:
                call_ms = int((time.monotonic() - t) * 1000)
                if status_of(e) == 403:
                    terminal = e
                    break
                raise
            await asyncio.sleep(2)

        check(
            "push eventually rejected with a terminal 403",
            terminal is not None,
            f"still accepted after {int(DEADLINE_S)}s",
        )
        check(
            "terminal code is storage_quota_exceeded",
            terminal is not None and getattr(terminal, "code", None) == "storage_quota_exceeded",
            str(getattr(terminal, "code", None)) if terminal else "n/a",
        )
        check(
            "terminal 403 surfaced immediately (not retried with backoff)",
            terminal is not None and call_ms < 2000,
            f"{call_ms}ms",
        )

        # consume must stay open while pushes are blocked: the rate-limit phase
        # left plenty of un-popped messages on this queue, so an empty result
        # here would mean the read path was blocked too.
        msgs = await q.queue(QUEUE).batch(1).wait(False).pop()
        check("consume still allowed while push-blocked", len(msgs) >= 1, f"{len(msgs)} msg(s)")
        if msgs:
            await q.ack(msgs, True)
    finally:
        await q.close()


# ---------------------------------------------------------------------------
# 4b. recovery once the driver clears the override
# ---------------------------------------------------------------------------
async def unblocked():
    q = client(HOST_A, KEY_A)
    deadline = time.monotonic() + DEADLINE_S
    released = False
    last_code = None
    try:
        while time.monotonic() < deadline and not released:
            try:
                r = await q.queue(QUEUE).push([{"release": time.time(), "run": RUN_ID}])
                released = bool(r) and r[0].get("status") == "queued"
            except httpx.HTTPStatusError as e:
                if status_of(e) != 403:
                    raise
                last_code = getattr(e, "code", None)
            if not released:
                await asyncio.sleep(2)
        check(
            "push accepted again after the override is cleared",
            released,
            f"still {last_code} after {int(DEADLINE_S)}s",
        )
    finally:
        await q.close()


PHASES = {
    "roundtrip": roundtrip,
    "isolation": isolation,
    "ratelimit": ratelimit,
    "blocked": blocked,
    "unblocked": unblocked,
}


def main():
    fn = PHASES.get(PHASE)
    if fn is None:
        print(
            f"sdk-smoke(py): unknown phase '{PHASE}' (want: {'|'.join(PHASES)})",
            file=sys.stderr,
        )
        sys.exit(2)
    try:
        asyncio.run(fn())
    except Exception as e:  # noqa: BLE001 -- a phase blowing up is a failure, not a crash
        bad(f"phase threw: {type(e).__name__}: {e} (status={status_of(e)} code={getattr(e, 'code', None)})")
    print(f"  -- py/{PHASE}: {STATE['pass']} ok, {STATE['fail']} fail")
    sys.exit(0 if STATE["fail"] == 0 else 1)


main()
